includes/class-wc-stripe-apple-pay-registration.php

```diff
@@ -188,16 +189,28 @@ public function update_domain_association_file( $force = false ) {
 				return true;
 			}
 
+			$error = null;
 			if ( ! file_exists( $path . '/' . $dir ) ) {
 				if ( ! @mkdir( $path . '/' . $dir, 0755 ) ) { // @codingStandardsIgnoreLine
-					WC_Stripe_Logger::log( 'Error: ' . __( 'Unable to create domain association folder to domain root.', 'woocommerce-gateway-stripe' ) );
-					return false;
+					$error = __( 'Unable to create domain association folder to domain root.', 'woocommerce-gateway-stripe' );
 				}
 			}
-
 			if ( ! @copy( WC_STRIPE_PLUGIN_PATH . '/' . $file, $fullpath ) ) { // @codingStandardsIgnoreLine
-				WC_Stripe_Logger::log( 'Error: ' . __( 'Unable to copy domain association file to domain root.', 'woocommerce-gateway-stripe' ) );
-				return false;
+				$error = __( 'Unable to copy domain association file to domain root.', 'woocommerce-gateway-stripe' );
+			}
+
+			if ( isset( $error ) ) {
+				$url            = 'https://' . $_SERVER['HTTP_HOST'] . '/' . $dir . '/' . $file;
+				$response       = wp_remote_get( $url );
+				$already_hosted = wp_remote_retrieve_body( $response ) === $new_contents;
+				if ( ! $already_hosted ) {
+					WC_Stripe_Logger::log(
+						'Error: ' . $error . ' ' .
+						/* translators: expected domain association file URL */
+						sprintf( __( 'To enable Apple Pay, domain association file must be hosted at %s.', 'woocommerce-gateway-stripe' ), $url )
+					);
+				}
+				return $already_hosted;
```

Comment:
I see how it might be good to verify that the file is actually being served, but seems sort of out of scope for this function? 

Is there a reason you decided to check that the file is served instead of just using [file_exists()](https://www.php.net/manual/en/function.file-exists.php)? I feel like the HTTP request might add non-trivial time to the function call, which we might not want.

Reply:
I probably should have laid out the rationale more clearly in the PR description – this is intended to avoid error reporting in cases where the file write fails (and so may or may not exist) but where the file is already being properly served anyway (such as with a rewrite rule on the web server, see https://github.com/woocommerce/woocommerce-gateway-stripe/issues/958).

I'm sure it does add time to the function call (by a factor of many orders of magnitude 😬) but only if the copy failed. It's unfortunate that this would run on every gateway settings page load, but I think that could be dealt with in a separate issue (see https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1330)

Reply:
Makes sense!